### PR TITLE
Add translation for boolean value in toggle columns for text cell

### PR DESCRIPTION
**Pull request by LouisDelbosc**
Related issue #334

### CODE REVIEW

**Comment by LouisDelbosc** [comment]
@paulfitz @berhalak  Is there a way to have translation in the tests ? I suspect the absence of translation is the root cause of the test failing.

**Comment by georgegevoian** [comment]
> @paulfitz @berhalak Is there a way to have translation in the tests ? I suspect the absence of translation is the root cause of the test failing.

Translations should already be included in browser tests.

I can replicate the failure from the test by trying to transform a numeric column of 0s and 1s to a toggle column. Sorry, haven't had time to look at the PR yet but I suspect it's trying to translate 0 and 1 (which are values of the right type) and failing because there are no keys for 0 or 1 in `*.locales.json`.

**Comment by LouisDelbosc** [comment]
> > @paulfitz @berhalak Is there a way to have translation in the tests ? I suspect the absence of translation is the root cause of the test failing.
> 
> Translations should already be included in browser tests.
> 
> I can replicate the failure from the test by trying to transform a numeric column of 0s and 1s to a toggle column. Sorry, haven't had time to look at the PR yet but I suspect it's trying to translate 0 and 1 (which are values of the right type) and failing because there are no keys for 0 or 1 in `*.locales.json`.

Adding translations for 0 and 1 fix the tests, thanks !

File: app/common/ValueFormatter.ts [hunk 1]
```diff
@@ -79,19 +79,25 @@ export class BaseFormatter {
    * Formats using this.format() if a value is of the right type for this formatter, or using
    * AnyFormatter otherwise. This method the recommended API. There is no need to override it.
    */
-  public formatAny(value: any): string {
-    return this.isRightType(value) ? this.format(value) : formatUnknown(value);
+  public formatAny(value: any, translate?: Function): string {
```
Comment (georgegevoian) [inline app/common/ValueFormatter.ts:82]:
```suggestion
  public formatAny(value: any, translate?: (val: string) => string): string {
```

File: app/common/ValueFormatter.ts [hunk 2]
```diff
@@ -79,19 +79,25 @@ export class BaseFormatter {
...
+    return this.isRightType(value) ? this.format(value, translate) : formatUnknown(value);
   }
 
   /**
    * Formats a value that matches the type of this formatter. This should be overridden by derived
    * classes to handle values in formatter-specific ways.
    */
-  protected format(value: any): string {
+  protected format(value: any, translate?: Function): string {
```
Comment (georgegevoian) [inline app/common/ValueFormatter.ts:90]:
```suggestion
  protected format(value: any, _translate?: (val: string) => string): string {
```

Sorry for nitpicking. This gets flagged as unused in some editors without the prefix.

File: app/common/ValueFormatter.ts [hunk 3]
```diff
@@ -79,19 +79,25 @@ export class BaseFormatter {
...
     return String(value);
   }
 }
 
+export class BoolFormatter extends BaseFormatter {
+  public format(value: boolean, translate?: (val: string) => string): string {
+    return translate ? translate(String(value)) : String(value);
```
Comment (georgegevoian) [inline app/common/ValueFormatter.ts:97]:
```suggestion
  public format(value: boolean | 0 | 1, translate?: (val: string) => string): string {
    if (typeof value === 'boolean' && translate) {
      return translate(String(value));
    }

    return super.format(value, translate);
```

What do you think of this approach? It should remove the need for creating translation keys for 0/1 (which aren't displayed, unlike true/false).

Comment (LouisDelbosc) [reply app/common/ValueFormatter.ts:97]: It's fine for me.

**Review by georgegevoian — APPROVED** [review]
Thanks @LouisDelbosc!

**Comment by yohanboniface** [comment]
Ooops ;)

[Screencast from 2023-01-10 16-30-39.webm](https://user-images.githubusercontent.com/146023/211593870-baec5b6b-ef90-4059-8789-0a673ccfc5a9.webm)

As you see, it seems that when trying to update the value (entering the field), we have to deal with the English version. I guess it's because this is the real value, but from a user point of view it's a bit surprising. Any idea how to fix/work around this ?

**Pull request merged**